spec/function.dd

```diff
@@ -173,124 +173,315 @@ int foo();
+pure int f2(int x) { ++a; return x; } // error, reads/writes mutable global
+pure int f3(int x) { return x + b; } // fine, reads immutable global
+---
+)
 
-        $(P A `pure` function accepting parameters with mutable indirections offers
-            what's called "weak purity" because it can change program state
-            transitively through its arguments. A `pure` function that has
-            no parameter with mutable indirections is called "strongly pure"
-            and fulfills the purity definition in traditional functional languages.
-            Weakly `pure` functions are useful as reusable building blocks for strongly pure functions.
-        )
+$(P A pure function cannot contain calls to functions that are not pure (Note: I/O functions
+such as $(REF writeln,std,stdio) are not pure.) Example:)
 
-        $(P To prevent mutation, D offers the `immutable` type qualifier.
-            If all of a `pure` function's parameters are `immutable` or
-            copied values without any indirections (e.g. `int`),
-            the type system guarantees no side effects.
-        )
-
-            ---
-            struct S { double x; }
-            pure int foo(immutable(int)[] arr, int num, S val)
-            {
-                //arr[num] = 1; // compile error
-                num = 2;        // has no side effect to the caller side
-                val.x = 3.14;   // ditto
-                return arr.length;
-            }
-            ---
+$(SPEC_RUNNABLE_EXAMPLE_FAIL
+---
+pure int f1(int x) { return x + 1; }
+pure int f2(int x) { return f1(x) - 1; } // fine
+extern int f3(int x);
+pure int f4(int x) { return true ? 42 : f3(x); } // error, calls impure function
+---
+)
 
-        $(P The maximum guarantee of `pure` is called "strong purity". It can enable
-            optimizations based on the fact
-            that a function is guaranteed to not mutate anything which isn't passed to it.
-            For cases where the compiler can guarantee that a pure function cannot
-            alter its arguments, it can enable full, functional purity (i.e. the guarantee
-            that the function will always return the same result for the same arguments).
-            To that end, a pure function:
-        )
+$(P If a function signature `f1` is covariant with a function signature `f2`, adding the `pure`
+attribute to `f1` preserves the covariance relationship. This implies that a pure method may
+override a non-pure method, but not vice versa. Example:)
 
-        $(UL
-        $(LI does not read or write any global or static mutable state)
-        $(LI cannot call functions that are not pure)
-        $(LI can override an impure function, but cannot be overridden by an impure function)
-        $(LI is covariant with an impure function)
-        $(LI cannot perform I/O)
-        )
+$(SPEC_RUNNABLE_EXAMPLE_COMPILE
+---
+class Base
+{
+    pure int fun(int x) { return x + 1; }
+    pure int gun(int x) { return x + 2; }
```

Comment:
shouldn't gun NOT be marked here?  Otherwise, Derived.gun has the exact same signature.

Reply:
Thanks, will fix.